src/cupbearer/data/backdoors.py

```diff
@@ -57,118 +66,187 @@ def __call__(self, sample: Tuple[np.ndarray, int]):
-            img = img + noise
-            img = np.clip(img, 0, 1)
-
-            target = self.target_class
-
-        return img, target
+        return img
 
 
 @dataclass
-class WanetBackdoor(Transform):
+class WanetBackdoor(Backdoor):
     """Implements trigger transform from "Wanet - Imperceptible Warping-based
     Backdoor Attack" by Anh Tuan Nguyen and Anh Tuan Tran, ICLR, 2021."""
 
-    p_backdoor: float = 1.0
-    p_noise: float = 0.0
-    control_grid_width: int = 4
-    warping_strength: float = 0.5
-    target_class: int = 0
-    grid_rescale: float = 1.0
+    p_noise: float = 0.0  # Probability of non-backdoor warping
+    control_grid_width: int = 4  # Side length of unscaled warping field
+    warping_strength: float = 0.5  # Strength of warping effect
+    grid_rescale: float = 1.0  # Factor to rescale grid from warping effect
+    _control_grid: Optional[
+        tuple[
+            list[list[float]],
+            list[list[float]],
+        ]
+    ] = None  # Used for reproducibility, typically not set manually
 
     def __post_init__(self):
+        super().__post_init__()
         self._warping_field = None
 
-        p_transform = self.p_backdoor + self.p_noise
-        assert 0 <= p_transform <= 1, "Probability must be between 0 and 1"
+        # Init control_grid so that it is saved in config
+        self.control_grid
+
+        assert 0 <= self.p_noise <= 1, "Probability must be between 0 and 1"
+        assert (
+            0 <= self.p_noise + self.p_backdoor <= 1
+        ), "Probability must be between 0 and 1"
+
+    @property
+    def control_grid(self) -> np.ndarray:
+        if self._control_grid is None:
+            control_grid_shape = (2, self.control_grid_width, self.control_grid_width)
+            control_grid = 2 * np.random.rand(*control_grid_shape) - 1
+            control_grid = control_grid / np.mean(np.abs(control_grid))
+            # N.B. the 0.5 comes from how the original did their rescaling, see
+            # https://github.com/ejnnr/cupbearer/pull/2#issuecomment-1688338610
+            control_grid = control_grid * 0.5 * self.warping_strength
+            self.control_grid = control_grid
+        else:
+            control_grid = np.array(self._control_grid)
+
+        control_grid_shape = (2, self.control_grid_width, self.control_grid_width)
+        assert control_grid.shape == control_grid_shape
+
+        return control_grid
+
+    @control_grid.setter
+    def control_grid(self, control_grid: np.ndarray):
+        control_grid_shape = (2, self.control_grid_width, self.control_grid_width)
+        if control_grid.shape != control_grid_shape:
+            raise ValueError("Control grid shape is incompatible.")
+
+        # We keep self._control_grid serializable
+        self._control_grid = tuple(control_grid.tolist())
+
+    @property
+    def warping_field(self) -> np.ndarray:
+        if self._warping_field is None:
+            raise AttributeError(
```

Comment:
Is there a reason not to just call `init_warping_field` here? If there are tricky cases where we might accidentally try to access `warping_field` before it makes sense to initialize, I'm all for failing loudly, but otherwise that might be simpler (e.g. means we can remove the `try ... except` clause in `__call__` below).

Reply:
`init_warping_field` needs the size of the input image. We could of course cache px, py instead to simplify __call__.

I have no preference, but I'm not sure if that alternative is any better. Let me know what you think.

Reply:
Oh sorry, I missed that, I agree keeping it the way it is sounds good then